git_machete/cmd.py

```diff
@@ -16,36 +17,42 @@
 
 # Core utils
 
+T = TypeVar('T')
+
+
 class MacheteException(Exception):
-    def __init__(self, msg, apply_fmt=True):
+    def __init__(self, msg: str, apply_fmt: bool = True) -> None:
         self.parameter = fmt(msg) if apply_fmt else msg
 
-    def __str__(self):
+    def __str__(self) -> str:
         return str(self.parameter)
 
 
-def excluding(iterable, s):
+def excluding(iterable: Iterable[T], s: Iterable[T]) -> List[T]:
     return list(filter(lambda x: x not in s, iterable))
 
 
-def flat_map(func, iterable):
+def flat_map(func: Callable[[T], List[T]], iterable: Iterable[T]) -> List[T]:
     return sum(map(func, iterable), [])
 
 
-def map_truthy_only(func, iterable):
+def map_truthy_only(func: Callable[[T], Optional[T]], iterable: Iterable[T]) -> List[T]:
     return list(filter(None, map(func, iterable)))
 
 
-def non_empty_lines(s):
+def non_empty_lines(s: str) -> List[str]:
     return list(filter(None, s.split("\n")))
 
 
 # Converts a lambda accepting N arguments to a lambda accepting one argument, an N-element tuple.
 # Name matching Scala's `tupled` on `FunctionX`.
-def tupled(f):
+def tupled(f: Callable[..., T]) -> Callable[..., T]:
```

Comment:
IIUC this could be a bit more specific:
```suggestion
def tupled(f: Callable[..., T]) -> Callable[Any, T]:
```
since the resulting lambda accepts exactly one arg

Reply:
Also... is there any umbrella type for `Tuple`, no matter the arity? like some `TupleN`?

Reply:
Or just `Tuple[...]`?

Reply:
For `Tuple[...]`:
```
git_machete/cmd.py:49: error: Unexpected '...'
git_machete/cmd.py:1489: error: Argument 1 to "tupled" has incompatible type "Callable[[str, str], str]"; expected "Callable[[Tuple[Any]], str]"
```
For `Tuple`:
```
git_machete/cmd.py:49: error: Missing type parameters for generic type "Tuple"
git_machete/cmd.py:1489: error: Argument 1 to "tupled" has incompatible type "Callable[[str, str], str]"; expected "Callable[[Tuple[Any, ...]], str]"
```

But the returned type can be more precise indeed